gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java

```diff
@@ -0,0 +1,203 @@
+  private Boolean isDone;
+  private String rewriteToken;
+  private BigInteger totalBytesRewritten;
+
+  private final StorageRpc storageRpc;
+
+  private BlobRewriter(Builder builder) {
+    this.serviceOptions = builder.serviceOptions;
+    this.source = builder.source;
+    this.sourceOptions = builder.sourceOptions;
+    this.target = builder.target;
+    this.targetOptions = builder.targetOptions;
+    this.blobSize = builder.blobSize;
+    this.isDone = builder.isDone;
+    this.rewriteToken = builder.rewriteToken;
+    this.totalBytesRewritten = firstNonNull(builder.totalBytesRewritten, BigInteger.ZERO);
+    this.maxBytesRewrittenPerCall = builder.maxBytesRewrittenPerCall;
+    this.storageRpc = serviceOptions.storageRpc();
+  }
+
+  static class Builder {
+
+    private final StorageOptions serviceOptions;
+    private final BlobId source;
+    private final Map<StorageRpc.Option, ?> sourceOptions;
+    private final BlobInfo target;
+    private final Map<StorageRpc.Option, ?> targetOptions;
+    private BigInteger blobSize;
+    private Boolean isDone;
+    private String rewriteToken;
+    private BigInteger totalBytesRewritten;
+    private Long maxBytesRewrittenPerCall;
+
+    Builder(StorageOptions serviceOptions, BlobId source, Map<StorageRpc.Option, ?> sourceOptions,
+        BlobInfo target, Map<StorageRpc.Option, ?> targetOptions) {
+      this.serviceOptions = serviceOptions;
+      this.source = source;
+      this.sourceOptions = sourceOptions;
+      this.target = target;
+      this.targetOptions = targetOptions;
+    }
+
+    Builder blobSize(BigInteger blobSize) {
+      this.blobSize = blobSize;
+      return this;
+    }
+
+    Builder isDone(Boolean isDone) {
+      this.isDone = isDone;
+      return this;
+    }
+
+    Builder rewriteToken(String rewriteToken) {
+      this.rewriteToken = rewriteToken;
+      return this;
+    }
+
+    Builder totalBytesRewritten(BigInteger totalBytesRewritten) {
+      this.totalBytesRewritten = totalBytesRewritten;
+      return this;
+    }
+
+    Builder maxBytesRewrittenPerCall(Long maxBytesRewrittenPerCall) {
+      this.maxBytesRewrittenPerCall = maxBytesRewrittenPerCall;
+      return this;
+    }
+
+    BlobRewriter build() {
+      return new BlobRewriter(this);
+    }
+  }
+
+  static Builder builder(StorageOptions options, BlobId source,
+      Map<StorageRpc.Option, ?> sourceOpt,
+      BlobInfo target, Map<StorageRpc.Option, ?> targetOpt) {
+    return new Builder(options, source, sourceOpt, target, targetOpt);
+  }
+
+  /**
+   * Returns the id of the source blob.
+   */
+  public BlobId source() {
+    return source;
+  }
+
+  /**
+   * Returns the info for the target blob. When {@link #isDone} is {@code true} this method returns
+   * the updated information for the just written blob.
+   */
+  public BlobInfo target() {
+    return target;
+  }
+
+  /**
+   * Size of the blob being copied. Returns {@code null} until the first copy request returns.
```

Comment:
I wrote it like this at first. With the first call to RPC in `writer`. I did not like that for rewrites that require only one RPC the `BlobRewriter.copyChunk` method was basically never used/unuseful. But I don't mind changing this.